source/fuzz/transformation_add_parameter.cpp

```diff
@@ -24,54 +24,113 @@ TransformationAddParameter::TransformationAddParameter(
+  message_.set_initializer_id_or_pointer_type_id(
+      initializer_id_or_pointer_type_id);
   message_.set_function_type_fresh_id(function_type_fresh_id);
 }
 
 bool TransformationAddParameter::IsApplicable(
     opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
-  // Check that function exists
+  // Check that function exists.
   const auto* function =
       fuzzerutil::FindFunction(ir_context, message_.function_id());
   if (!function ||
       fuzzerutil::FunctionIsEntryPoint(ir_context, function->result_id())) {
     return false;
   }
 
-  // Check that |initializer_id| is valid.
-  const auto* initializer_inst =
-      ir_context->get_def_use_mgr()->GetDef(message_.initializer_id());
-
+  // Check that |initializer_id| is valid. Consider also the case where
+  // |initalizer_id| is a type id of a pointer.
+  const auto* initializer_inst = ir_context->get_def_use_mgr()->GetDef(
+      message_.initializer_id_or_pointer_type_id());
+  opt::analysis::Type* initializer_type;
   if (!initializer_inst) {
     return false;
   }
+  auto type = ir_context->get_type_mgr()->GetType(
+      message_.initializer_id_or_pointer_type_id());
+  if (type && !type->AsPointer()) {
+    return false;
+  }
+  if (type && type->AsPointer()) {
+    initializer_type = type;
+  } else {
+    initializer_type =
+        ir_context->get_type_mgr()->GetType(initializer_inst->type_id());
+  }
 
   // Check that initializer's type is valid.
-  const auto* initializer_type =
-      ir_context->get_type_mgr()->GetType(initializer_inst->type_id());
-
   if (!initializer_type || !IsParameterTypeSupported(*initializer_type)) {
     return false;
   }
 
+  // Consider pointer types separately.
+  bool is_valid = true;
+  if (initializer_type->kind() == opt::analysis::Type::kPointer) {
+    uint32_t pointer_type_id = message_.initializer_id_or_pointer_type_id();
+    auto storage_class =
+        fuzzerutil::GetStorageClassFromPointerType(ir_context, pointer_type_id);
+    switch (storage_class) {
+      case SpvStorageClassFunction:
+        for (auto* instr :
+             fuzzerutil::GetCallers(ir_context, message_.function_id())) {
+          auto block = ir_context->get_instr_block(instr);
+          auto function_id = block->GetParent()->result_id();
+          // If there is no available local variable in at least one caller,
+          // the transformation is invalid.
+          if (!fuzzerutil::MaybeGetLocalVariable(ir_context, pointer_type_id,
+                                                 function_id)) {
+            is_valid = false;
+            break;
+          }
+        }
+        break;
+      case SpvStorageClassPrivate:
+      case SpvStorageClassWorkgroup:
+        // If there is no available global variable, the transformation is
+        // invalid.
+        if (!fuzzerutil::MaybeGetGlobalVariable(ir_context, pointer_type_id)) {
+          is_valid = false;
+        }
+        break;
+      default:
+        break;
```

Comment:
It would be good to make an assert here saying that storage class is not supported.